caas/kubernetes/provider/k8s.go

```diff
@@ -1327,102 +1326,94 @@ func (k *kubernetesClient) deleteAllPods(appName, deploymentName string) error {
-		}
-		if mountPath == "" {
-			mountPath = fmt.Sprintf("%s/fs/%s/%s/%d", baseDir, appName, fs.StorageName, i)
-		}
-		fsSize, err := resource.ParseQuantity(fmt.Sprintf("%dMi", fs.Size))
-		if err != nil {
-			return errors.Annotatef(err, "invalid volume size %v", fs.Size)
-		}
-
-		var volumeSource *core.VolumeSource
-		switch fs.Provider {
-		case K8s_ProviderType:
-		case provider.RootfsProviderType:
-			volumeSource = &core.VolumeSource{
-				EmptyDir: &core.EmptyDirVolumeSource{
-					SizeLimit: &fsSize,
-				},
-			}
-		case provider.TmpfsProviderType:
-			medium, ok := fs.Attributes[storageMedium]
-			if !ok {
-				medium = core.StorageMediumMemory
-			}
-			volumeSource = &core.VolumeSource{
-				EmptyDir: &core.EmptyDirVolumeSource{
-					Medium:    core.StorageMedium(fmt.Sprintf("%v", medium)),
-					SizeLimit: &fsSize,
-				},
-			}
-		default:
-			return errors.NotValidf("charm storage provider type %q for %v", fs.Provider, fs.StorageName)
-		}
-		if volumeSource != nil {
-			logger.Debugf("using emptyDir for %s filesystem %s", appName, fs.StorageName)
-			volName := fmt.Sprintf("%s-%d", fs.StorageName, i)
-			podSpec.Containers[0].VolumeMounts = append(podSpec.Containers[0].VolumeMounts, core.VolumeMount{
-				Name:      volName,
-				MountPath: mountPath,
-			})
-			podSpec.Volumes = append(podSpec.Volumes, core.Volume{
-				Name:         volName,
-				VolumeSource: *volumeSource,
-			})
-			continue
-		}
 
-		pvcNamePrefix := fmt.Sprintf("%s-%s", fs.StorageName, randPrefix)
-		if legacy {
-			pvcNamePrefix = fmt.Sprintf("juju-%s-%d", fs.StorageName, i)
+	var volumeSource *core.VolumeSource
+	switch fs.Provider {
+	case K8s_ProviderType:
+	case provider.RootfsProviderType:
+		volumeSource = &core.VolumeSource{
+			EmptyDir: &core.EmptyDirVolumeSource{
+				SizeLimit: &fsSize,
+			},
 		}
-		params := volumeParams{
-			pvcName:             pvcNamePrefix,
-			requestedVolumeSize: fsSize,
+	case provider.TmpfsProviderType:
+		medium, ok := fs.Attributes[storageMedium]
+		if !ok {
+			medium = core.StorageMediumMemory
 		}
-		params.storageConfig, err = newStorageConfig(fs.Attributes)
-		if err != nil {
-			return errors.Annotatef(err, "invalid storage configuration for %v", fs.StorageName)
+		volumeSource = &core.VolumeSource{
+			EmptyDir: &core.EmptyDirVolumeSource{
+				Medium:    core.StorageMedium(fmt.Sprintf("%v", medium)),
+				SizeLimit: &fsSize,
+			},
 		}
-		pvcSpec, err := k.maybeGetVolumeClaimSpec(params)
-		if err != nil {
-			return errors.Annotatef(err, "finding volume for %s", fs.StorageName)
+	default:
+		return nil, nil, errors.NotValidf("charm storage provider type %q for %v", fs.Provider, fs.StorageName)
+	}
+	if volumeSource != nil {
+		volName := fmt.Sprintf("%s-%d", fs.StorageName, i)
+		vol = &core.Volume{
+			Name:         volName,
+			VolumeSource: *volumeSource,
 		}
+		return
+	}
+	params := volumeParams{
+		pvcName:             pvcNameGetter(i, fs.StorageName),
+		requestedVolumeSize: fsSize,
```

Comment:
Need to update storage.KubernetesFilesystemParams to have a core.PersistentVolumeAccessMode field, then pass it into volumeParams.accessMode.

Without this, deployments and daemonsets won't work past a scale of 1

Reply:
good point!
if scale == 1, RWO would work.
but if scale > 1, it has to be either ROM or RWM - these two have very limited support for different cloud.

https://kubernetes.io/docs/concepts/storage/persistent-volumes/#access-modes

Reply:
As discussed, we will solve this in the next PR https://trello.com/c/adQGDuri